File: java/VendingMachine/test/com/ravendarque/InventoryShould.java
package com.ravendarque;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class InventoryShould {

    @Test
    void ProvideEmptyInventoryListWhenNoItemsInInventory() {

        Inventory inventory = new Inventory();
        List<Item> actualInventoryList = inventory.getItems();

        assertTrue(actualInventoryList.isEmpty());
    }

    @Test
    void ProvideOneItemInInventoryListWhenOneItemIsInInventory() {

        String dummyName = "Test item";
        double dummyPrice = 0;
        Item testItem = new Item(dummyName, dummyPrice);

        List<Item> expectedInventoryList = new ArrayList<>();
        expectedInventoryList.add(testItem);

        Inventory inventory = new Inventory();
        inventory.addItem(testItem);
        List<Item> actualInventoryList = inventory.getItems();

        assertEquals(expectedInventoryList, actualInventoryList);
    }

    @Test
    void ProvideEmptyInventoryListWhenOneItemIsAddedThenRemoved() {

        String dummyName = "Test item";
        double dummyPrice = 0;
        Item testItem = new Item(dummyName, dummyPrice);

        Inventory inventory = new Inventory();
        inventory.addItem(testItem);
        inventory.removeItem(testItem);

        List<Item> actualInventoryList = inventory.getItems();

        assertTrue(actualInventoryList.isEmpty());
    }

    @Test
    void ReturnOnlyDrinkItemsWhenQueriedForDrinkItems() {


        String testDrinkItemDisplayName = "Test drink item";
        double dummyDrinkItemPrice = 0;
        Drink testDrinkItem = new Drink(testDrinkItemDisplayName, dummyDrinkItemPrice);

        String testOtherItemDisplayName = "Test other item";
        double dummyOtherItemPrice = 0;
        Item testOtherItem = new Item(testOtherItemDisplayName, dummyOtherItemPrice);

        List<Item> expectedInventoryList = new ArrayList<>();
        expectedInventoryList.add(testDrinkItem);

        Inventory testInventory = new Inventory();
        testInventory.addItem(testDrinkItem);
        testInventory.addItem(testOtherItem);

        List<Item> actualInventoryList = testInventory.getItemsByType(Drink.class);

        assertEquals(expectedInventoryList, actualInventoryList);
    }
}

File: C++/VendingMachine/VendingMachine.UnitTestsG/InventoryShould.cpp
#include "pch.h"
#include "../VendingMachine/Inventory.h"

TEST(InventoryShould, ProvideEmptyInventoryListWhenNoItemsInInventory)
{
	const Inventory test_inventory;

	const auto actual_inventory_list = test_inventory.GetItems();

	EXPECT_TRUE(actual_inventory_list.empty());
}

TEST(InventoryShould, ProvideOneItemInInventoryListWhenOneItemIsInInventory)
{
	const std::string dummy_display_name = "Test Item";
	const double dummy_price = 1;

	Item test_item(dummy_display_name, dummy_price);

	Inventory inventory;
	inventory.AddItem(test_item);
	const auto actual_inventory_list = inventory.GetItems();

	EXPECT_EQ(actual_inventory_list.size(), 1);
	EXPECT_EQ(actual_inventory_list[0], &test_item);
}

TEST(InventoryShould, ProvideEmptyInventoryListWhenOneItemIsAddedThenRemoved)
{
	const std::string dummy_display_name = "Test Item";
	const double dummy_price = 1;

	Item test_item(dummy_display_name, dummy_price);

	Inventory inventory;
	inventory.AddItem(test_item);
	inventory.RemoveItem(test_item);
	const auto actual_inventory_list = inventory.GetItems();

	EXPECT_TRUE(actual_inventory_list.empty());
}
